crates/oxc_linter/src/rules/typescript/no_unused_expressions.rs

```diff
@@ -0,0 +1,416 @@
+use oxc_ast::{
+    ast::{ChainElement, Expression, UnaryOperator},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use serde_json::Value;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_unused_expressions_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Disallow unused expressions")
+        .with_help("Consider removing this expression")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUnusedExpressions(Box<NoUnusedExpressionsConfig>);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUnusedExpressionsConfig {
+    allow_short_circuit: bool,
+    allow_ternary: bool,
+    allow_tagged_templates: bool,
+    enforce_for_jsx: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// This rule disallows unused expressions.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Unused expressions are usually a mistake. They can be a symptom of a bug or a misunderstanding of the code.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```ts
+    /// Set<number>;
+    /// 1 as number;
+    /// window!;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```ts
+    /// const foo = new Set<number>();
+    /// ```
+    NoUnusedExpressions,
+    restriction
```

Comment:
This rule is interesting, I may want to test it in ecosystem ci and see if we can make this `correctness`.